weblab/repocache/models.py

```diff
@@ -0,0 +1,114 @@
+    and should not be changed without first changing the repo.
+    """
+    entity = models.OneToOneField(Entity)
+
+    @property
+    def visibility(self):
+        """
+        Visibility of the entity (this is based on the visibility of the latest
+        version)
+
+        :return: string representing visibility, or PRIVATE if no versions found
+        """
+        try:
+            return self.latest_version.visibility
+        except ObjectDoesNotExist:
+            return Visibility.PRIVATE
+
+    @property
+    def latest_version(self):
+        return self.versions.latest()
+
+    def get_version(self, sha):
+        """
+        Get a version of the entity
+
+        :param sha: hex string of the commit SHA of the version
+
+        :return: CachedEntityVersion object
+        :raise: RepoCacheMiss if entity does not exist in cache, or has no versions
+        """
+        try:
+            if sha == 'latest':
+                return self.latest_version
+            else:
+                return self.versions.get(sha=sha)
+        except ObjectDoesNotExist:
+            raise RepoCacheMiss("Entity version not found")
+
+    def add_version(self, sha):
+        """
+        Add an entity version to the cache
+
+        :param sha: hex string of the commit SHA of the version
+        :return CachedEntityVersion object
+        """
+        commit = self.entity.repo.get_commit(sha)
+        visibility = self.entity.get_visibility_from_repo(commit)
+        return CachedEntityVersion.objects.create(
+            entity=self,
+            sha=commit.hexsha,
+            timestamp=commit.committed_at,
+            visibility=visibility,
+        )
+
+
+class CachedEntityVersion(VisibilityModelMixin):
+    """
+    Cache for a single version / commit in an entity's repository
+    """
+    entity = models.ForeignKey(CachedEntity, related_name='versions')
+    sha = models.CharField(max_length=40)
+    timestamp = models.DateTimeField()
+
+    class Meta:
+        unique_together = ['entity', 'sha']
+        get_latest_by = 'timestamp'
+
+    def set_visibility(self, visibility):
+        """
+        Set the visibility of this version, if it exists
+
+        :param visibility: string representing visibility
+        """
+        self.visibility = visibility
+        self.save()
+
+    def tag(self, tagname):
+        """
+        Add a tag for this version
+
+        :param tagname: Tag name
+        """
+        return CachedEntityTag.objects.create(
+            entity=self.entity,
+            version=self,
+            tag=tagname,
+        )
+
+
+class CachedEntityTag(models.Model):
+    """
+    Cache for a tag in an entity's repository
+    """
+    entity = models.ForeignKey(CachedEntity, related_name='tags')
+    tag = models.CharField(max_length=255)
+    version = models.ForeignKey(CachedEntityVersion, related_name='tags')
+
+    class Meta:
+        unique_together = ['entity', 'tag']
```

Comment:
Does this automatically handle the case where a tag is updated to point at a newer version?

Reply:
I don't think we have this functionality in the webapp, but this should be easily done if it is added. (I'm not sure if I've understood this right)

Reply:
Good point, we don't have the facility to delete tags (yet).